File: experiments/btc-updown/src/btc_updown_ml/artifact.py
from __future__ import annotations

from dataclasses import asdict, dataclass
import hashlib
import json
import math
import struct
from typing import Any, Iterable, Mapping


ARTIFACT_VERSION = "linear_logit_artifact_v1"
VALID_TASKS = {
    "settlement_probability_residual",
    "fok_fill_probability",
    "fok_post_fill_toxicity_probability",
}


@dataclass(frozen=True)
class LinearFeatureTransform:
    name: str
    mean: float
    scale: float
    coefficient: float

    def validate(self) -> None:
        if not self.name.strip():
            raise ValueError("feature transform name must not be empty")
        if not all(math.isfinite(value) for value in (self.mean, self.scale, self.coefficient)):
            raise ValueError(f"feature transform {self.name!r} contains a non-finite value")
        if self.scale <= 0.0:
            raise ValueError(f"feature transform {self.name!r} scale must be positive")


@dataclass(frozen=True)
class LinearLogitArtifact:
    artifact_version: str
    model_version: str
    task: str
    prior_kind: str
    feature_schema_version: str
    feature_schema_sha256: str
    dataset_manifest_sha256: str | None
    trained_through_ms: int | None
    intercept: float
    features: tuple[LinearFeatureTransform, ...]
    artifact_sha256: str

    def validate(self) -> None:
        if self.artifact_version != ARTIFACT_VERSION:
            raise ValueError(f"unsupported artifact version {self.artifact_version!r}")
        if not self.model_version.strip() or not self.feature_schema_version.strip():
            raise ValueError("model and feature schema versions must not be empty")
        if self.task not in VALID_TASKS:
            raise ValueError(f"unsupported ML task {self.task!r}")
        if self.prior_kind != "provided_probability":
            raise ValueError(f"unsupported prior kind {self.prior_kind!r}")
        if not math.isfinite(self.intercept):
            raise ValueError("intercept must be finite")
        if not self.features:
            raise ValueError("artifact features must not be empty")
        names: set[str] = set()
        for feature in self.features:
            feature.validate()
            if feature.name in names:
                raise ValueError(f"duplicate artifact feature {feature.name!r}")
            names.add(feature.name)
        expected_schema_hash = hashlib.sha256(
            canonical_feature_schema(
                self.feature_schema_version, [feature.name for feature in self.features]
            ).encode("utf-8")
        ).hexdigest()
        if self.feature_schema_sha256 != expected_schema_hash:
            raise ValueError("feature schema hash mismatch")
        if self.dataset_manifest_sha256 is not None and not _is_sha256(
            self.dataset_manifest_sha256
        ):
            raise ValueError("dataset manifest hash must be a SHA-256 hex digest")
        expected_artifact_hash = hashlib.sha256(
            canonical_artifact(self).encode("utf-8")
        ).hexdigest()
        if self.artifact_sha256 != expected_artifact_hash:
            raise ValueError("artifact hash mismatch")

    def to_mapping(self) -> dict[str, Any]:
        value = asdict(self)
        value["features"] = [asdict(feature) for feature in self.features]
        return value

    def to_json(self) -> str:
        self.validate()
        return json.dumps(self.to_mapping(), indent=2, sort_keys=True, allow_nan=False)

    @classmethod
    def from_mapping(cls, value: Mapping[str, Any]) -> "LinearLogitArtifact":
        artifact = cls(
            artifact_version=str(value["artifact_version"]),
            model_version=str(value["model_version"]),
            task=str(value["task"]),
            prior_kind=str(value["prior_kind"]),
            feature_schema_version=str(value["feature_schema_version"]),
            feature_schema_sha256=str(value["feature_schema_sha256"]),
            dataset_manifest_sha256=(
                None
                if value.get("dataset_manifest_sha256") is None
                else str(value["dataset_manifest_sha256"])
            ),
            trained_through_ms=(
                None if value.get("trained_through_ms") is None else int(value["trained_through_ms"])
            ),
            intercept=float(value["intercept"]),
            features=tuple(
                LinearFeatureTransform(
                    name=str(feature["name"]),
                    mean=float(feature["mean"]),
                    scale=float(feature["scale"]),
                    coefficient=float(feature["coefficient"]),
                )
                for feature in value["features"]
            ),
            artifact_sha256=str(value["artifact_sha256"]),
        )
        artifact.validate()
        return artifact


def build_artifact(
    *,
    model_version: str,
    task: str,
    feature_schema_version: str,
    features: Iterable[LinearFeatureTransform],
    intercept: float,
    dataset_manifest_sha256: str | None,
    trained_through_ms: int | None,
) -> LinearLogitArtifact:
    feature_tuple = tuple(features)
    schema_hash = hashlib.sha256(
        canonical_feature_schema(
            feature_schema_version, [feature.name for feature in feature_tuple]
        ).encode("utf-8")
    ).hexdigest()
    unsigned = LinearLogitArtifact(
        artifact_version=ARTIFACT_VERSION,
        model_version=model_version,
        task=task,
        prior_kind="provided_probability",
        feature_schema_version=feature_schema_version,
        feature_schema_sha256=schema_hash,
        dataset_manifest_sha256=dataset_manifest_sha256,
        trained_through_ms=trained_through_ms,
        intercept=intercept,
        features=feature_tuple,
        artifact_sha256="",
    )
    artifact = LinearLogitArtifact(
        **{**unsigned.to_mapping(), "features": feature_tuple, "artifact_sha256": hashlib.sha256(
            canonical_artifact(unsigned).encode("utf-8")
        ).hexdigest()}
    )
    artifact.validate()
    return artifact


def schema_canary_v0(
    *, task: str, feature_schema_version: str, feature_names: Iterable[str]
) -> LinearLogitArtifact:
    """Return a non-alpha artifact that exactly reproduces its supplied prior."""
    return build_artifact(
        model_version="schema_canary_v0",
        task=task,
        feature_schema_version=feature_schema_version,
        features=(
            LinearFeatureTransform(name=name, mean=0.0, scale=1.0, coefficient=0.0)
            for name in feature_names
        ),
        intercept=0.0,
        dataset_manifest_sha256=None,
        trained_through_ms=None,
    )


def canonical_feature_schema(schema_version: str, feature_names: Iterable[str]) -> str:
    names = tuple(feature_names)
    output = "ml_feature_schema_v1\n"
    output += _text("schema_version", schema_version)
    output += f"feature_count={len(names)}\n"
    for index, name in enumerate(names):
        output += _text(f"feature[{index}]", name)
    return output


def canonical_artifact(artifact: LinearLogitArtifact) -> str:
    output = "linear_logit_artifact_canonical_v1\n"
    output += _text("artifact_version", artifact.artifact_version)
    output += _text("model_version", artifact.model_version)
    output += _text("task", artifact.task)
    output += _text("prior_kind", artifact.prior_kind)
    output += _text("feature_schema_version", artifact.feature_schema_version)
    output += _text("feature_schema_sha256", artifact.feature_schema_sha256)
    output += _optional_text("dataset_manifest_sha256", artifact.dataset_manifest_sha256)
    output += _optional_i64("trained_through_ms", artifact.trained_through_ms)
    output += _float("intercept", artifact.intercept)
    output += f"feature_count={len(artifact.features)}\n"
    for index, feature in enumerate(artifact.features):
        output += _text(f"feature[{index}].name", feature.name)
        output += _float(f"feature[{index}].mean", feature.mean)
        output += _float(f"feature[{index}].scale", feature.scale)
        output += _float(f"feature[{index}].coefficient", feature.coefficient)
    return output


def _text(key: str, value: str) -> str:
    return f"{key}.utf8_bytes={len(value.encode('utf-8'))}:{value}\n"


def _float(key: str, value: float) -> str:
    bits = struct.unpack(">Q", struct.pack(">d", value))[0]
    return f"{key}.f64_bits={bits:016x}\n"


def _optional_text(key: str, value: str | None) -> str:
    if value is None:
        return f"{key}.present=0\n"
    return f"{key}.present=1\n" + _text(key, value)


def _optional_i64(key: str, value: int | None) -> str:
    if value is None:
        return f"{key}.present=0\n"
    return f"{key}.present=1\n{key}.value={value}\n"


def _is_sha256(value: str) -> bool:
    return len(value) == 64 and all(character in "0123456789abcdefABCDEF" for character in value)

File: packages/polymarket-bot/src/ml/artifact.rs
use std::{collections::HashSet, error::Error, fmt};

use serde::{Deserialize, Serialize};

use super::features::{canonical_feature_schema, push_float, push_text, sha256_hex};

pub const LINEAR_LOGIT_ARTIFACT_VERSION: &str = "linear_logit_artifact_v1";

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum MlTask {
    SettlementProbabilityResidual,
    FokFillProbability,
    FokPostFillToxicityProbability,
}

impl MlTask {
    pub fn as_str(self) -> &'static str {
        match self {
            Self::SettlementProbabilityResidual => "settlement_probability_residual",
            Self::FokFillProbability => "fok_fill_probability",
            Self::FokPostFillToxicityProbability => "fok_post_fill_toxicity_probability",
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum PriorKind {
    ProvidedProbability,
}

impl PriorKind {
    fn as_str(self) -> &'static str {
        match self {
            Self::ProvidedProbability => "provided_probability",
        }
    }
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct LinearFeatureTransform {
    name: String,
    mean: f64,
    scale: f64,
    coefficient: f64,
}

impl LinearFeatureTransform {
    pub fn new(
        name: impl Into<String>,
        mean: f64,
        scale: f64,
        coefficient: f64,
    ) -> Result<Self, ArtifactError> {
        let transform = Self {
            name: name.into(),
            mean,
            scale,
            coefficient,
        };
        transform.validate()?;
        Ok(transform)
    }

    pub fn name(&self) -> &str {
        &self.name
    }

    pub fn mean(&self) -> f64 {
        self.mean
    }

    pub fn scale(&self) -> f64 {
        self.scale
    }

    pub fn coefficient(&self) -> f64 {
        self.coefficient
    }

    fn validate(&self) -> Result<(), ArtifactError> {
        if self.name.trim().is_empty() {
            return Err(ArtifactError::EmptyFeatureName);
        }
        if !self.mean.is_finite() {
            return Err(ArtifactError::NonFiniteParameter(format!(
                "{}.mean",
                self.name
            )));
        }
        if !self.scale.is_finite() || self.scale <= 0.0 {
            return Err(ArtifactError::InvalidScale(self.name.clone()));
        }
        if !self.coefficient.is_finite() {
            return Err(ArtifactError::NonFiniteParameter(format!(
                "{}.coefficient",
                self.name
            )));
        }
        Ok(())
    }
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct LinearLogitArtifact {
    artifact_version: String,
    model_version: String,
    task: MlTask,
    prior_kind: PriorKind,
    feature_schema_version: String,
    feature_schema_sha256: String,
    dataset_manifest_sha256: Option<String>,
    trained_through_ms: Option<i64>,
    intercept: f64,
    features: Vec<LinearFeatureTransform>,
    artifact_sha256: String,
}

impl LinearLogitArtifact {
    #[allow(clippy::too_many_arguments)]
    pub fn new(
        model_version: impl Into<String>,
        task: MlTask,
        prior_kind: PriorKind,
        feature_schema_version: impl Into<String>,
        dataset_manifest_sha256: Option<String>,
        trained_through_ms: Option<i64>,
        intercept: f64,
        features: Vec<LinearFeatureTransform>,
    ) -> Result<Self, ArtifactError> {
        let feature_schema_version = feature_schema_version.into();
        let names = features.iter().map(|feature| feature.name());
        let feature_schema_sha256 =
            sha256_hex(&canonical_feature_schema(&feature_schema_version, names));
        let mut artifact = Self {
            artifact_version: LINEAR_LOGIT_ARTIFACT_VERSION.to_string(),
            model_version: model_version.into(),
            task,
            prior_kind,
            feature_schema_version,
            feature_schema_sha256,
            dataset_manifest_sha256,
            trained_through_ms,
            intercept,
            features,
            artifact_sha256: String::new(),
        };
        artifact.validate_content()?;
        artifact.artifact_sha256 = sha256_hex(&canonical_artifact(&artifact));
        Ok(artifact)
    }

    /// A pipeline-validation artifact. It contains no learned alpha and exactly reproduces the
    /// provided prior because every residual coefficient and the intercept are zero.
    pub fn schema_canary_v0<'a>(
        task: MlTask,
        feature_schema_version: impl Into<String>,
        feature_names: impl IntoIterator<Item = &'a str>,
    ) -> Result<Self, ArtifactError> {
        let features = feature_names
            .into_iter()
            .map(|name| LinearFeatureTransform::new(name, 0.0, 1.0, 0.0))
            .collect::<Result<Vec<_>, _>>()?;
        Self::new(
            "schema_canary_v0",
            task,
            PriorKind::ProvidedProbability,
            feature_schema_version,
            None,
            None,
            0.0,
            features,
        )
    }

    pub fn from_json(value: &str) -> Result<Self, ArtifactError> {
        let artifact: Self = serde_json::from_str(value)
            .map_err(|error| ArtifactError::InvalidJson(error.to_string()))?;
        artifact.validate()?;
        Ok(artifact)
    }

    pub fn to_json(&self) -> Result<String, ArtifactError> {
        self.validate()?;
        serde_json::to_string_pretty(self)
            .map_err(|error| ArtifactError::InvalidJson(error.to_string()))
    }

    pub fn validate(&self) -> Result<(), ArtifactError> {
        self.validate_content()?;
        let expected = sha256_hex(&canonical_artifact(self));
        if expected != self.artifact_sha256 {
            return Err(ArtifactError::ArtifactHashMismatch {
                expected,
                actual: self.artifact_sha256.clone(),
            });
        }
        Ok(())
    }

    fn validate_content(&self) -> Result<(), ArtifactError> {
        if self.artifact_version != LINEAR_LOGIT_ARTIFACT_VERSION {
            return Err(ArtifactError::UnsupportedArtifactVersion(
                self.artifact_version.clone(),
            ));
        }
        if self.model_version.trim().is_empty() {
            return Err(ArtifactError::EmptyModelVersion);
        }
        if self.feature_schema_version.trim().is_empty() {
            return Err(ArtifactError::EmptyFeatureSchemaVersion);
        }
        if !self.intercept.is_finite() {
            return Err(ArtifactError::NonFiniteParameter("intercept".to_string()));
        }
        if self.features.is_empty() {
            return Err(ArtifactError::EmptyFeatureList);
        }
        let mut names = HashSet::with_capacity(self.features.len());
        for feature in &self.features {
            feature.validate()?;
            if !names.insert(feature.name().to_string()) {
                return Err(ArtifactError::DuplicateFeatureName(
                    feature.name().to_string(),
                ));
            }
        }
        let schema = sha256_hex(&canonical_feature_schema(
            &self.feature_schema_version,
            self.features.iter().map(|feature| feature.name()),
        ));
        if schema != self.feature_schema_sha256 {
            return Err(ArtifactError::FeatureSchemaHashMismatch {
                expected: schema,
                actual: self.feature_schema_sha256.clone(),
            });
        }
        if let Some(hash) = &self.dataset_manifest_sha256 {
            if !is_sha256_hex(hash) {
                return Err(ArtifactError::InvalidDatasetManifestHash(hash.clone()));
            }
        }
        Ok(())
    }

    pub fn model_version(&self) -> &str {
        &self.model_version
    }

    pub fn task(&self) -> MlTask {
        self.task
    }

    pub fn prior_kind(&self) -> PriorKind {
        self.prior_kind
    }

    pub fn feature_schema_version(&self) -> &str {
        &self.feature_schema_version
    }

    pub fn feature_schema_sha256(&self) -> &str {
        &self.feature_schema_sha256
    }

    pub fn dataset_manifest_sha256(&self) -> Option<&str> {
        self.dataset_manifest_sha256.as_deref()
    }

    pub fn intercept(&self) -> f64 {
        self.intercept
    }

    pub fn features(&self) -> &[LinearFeatureTransform] {
        &self.features
    }

    pub fn artifact_sha256(&self) -> &str {
        &self.artifact_sha256
    }
}

pub fn canonical_artifact(artifact: &LinearLogitArtifact) -> String {
    let mut output = String::from("linear_logit_artifact_canonical_v1\n");
    push_text(&mut output, "artifact_version", &artifact.artifact_version);
    push_text(&mut output, "model_version", &artifact.model_version);
    push_text(&mut output, "task", artifact.task.as_str());
    push_text(&mut output, "prior_kind", artifact.prior_kind.as_str());
    push_text(
        &mut output,
        "feature_schema_version",
        &artifact.feature_schema_version,
    );
    push_text(
        &mut output,
        "feature_schema_sha256",
        &artifact.feature_schema_sha256,
    );
    push_optional_text(
        &mut output,
        "dataset_manifest_sha256",
        artifact.dataset_manifest_sha256.as_deref(),
    );
    push_optional_i64(
        &mut output,
        "trained_through_ms",
        artifact.trained_through_ms,
    );
    push_float(&mut output, "intercept", artifact.intercept);
    output.push_str(&format!("feature_count={}\n", artifact.features.len()));
    for (index, feature) in artifact.features.iter().enumerate() {
        push_text(
            &mut output,
            &format!("feature[{index}].name"),
            feature.name(),
        );
        push_float(
            &mut output,
            &format!("feature[{index}].mean"),
            feature.mean(),
        );
        push_float(
            &mut output,
            &format!("feature[{index}].scale"),
            feature.scale(),
        );
        push_float(
            &mut output,
            &format!("feature[{index}].coefficient"),
            feature.coefficient(),
        );
    }
    output
}

fn push_optional_text(output: &mut String, key: &str, value: Option<&str>) {
    match value {
        Some(value) => {
            output.push_str(&format!("{key}.present=1\n"));
            push_text(output, key, value);
        }
        None => output.push_str(&format!("{key}.present=0\n")),
    }
}

fn push_optional_i64(output: &mut String, key: &str, value: Option<i64>) {
    match value {
        Some(value) => output.push_str(&format!("{key}.present=1\n{key}.value={value}\n")),
        None => output.push_str(&format!("{key}.present=0\n")),
    }
}

fn is_sha256_hex(value: &str) -> bool {
    value.len() == 64 && value.chars().all(|character| character.is_ascii_hexdigit())
}

#[derive(Debug, Clone, PartialEq)]
pub enum ArtifactError {
    InvalidJson(String),
    UnsupportedArtifactVersion(String),
    EmptyModelVersion,
    EmptyFeatureSchemaVersion,
    EmptyFeatureName,
    EmptyFeatureList,
    DuplicateFeatureName(String),
    NonFiniteParameter(String),
    InvalidScale(String),
    InvalidDatasetManifestHash(String),
    FeatureSchemaHashMismatch { expected: String, actual: String },
    ArtifactHashMismatch { expected: String, actual: String },
}

impl fmt::Display for ArtifactError {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(formatter, "{self:?}")
    }
}

impl Error for ArtifactError {}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn artifact_round_trip_preserves_and_validates_hash() {
        let artifact = LinearLogitArtifact::schema_canary_v0(
            MlTask::SettlementProbabilityResidual,
            "features-v1",
            ["distance", "seconds"],
        )
        .unwrap();
        // This fixture is shared with the standard-library Python implementation.
        assert_eq!(
            artifact.feature_schema_sha256(),
            "1502f8f90e7104b04834bbc01ab9675b16a70465e2c01eaadf0468fbd720f258"
        );
        assert_eq!(
            artifact.artifact_sha256(),
            "82c212a8bb1ce1705b6b643ecdffc0f5b13143549b71a88184498d08190f5aef"
        );
        let json = artifact.to_json().unwrap();
        let decoded = LinearLogitArtifact::from_json(&json).unwrap();
        assert_eq!(decoded, artifact);
    }

    #[test]
    fn tampered_json_is_rejected() {
        let artifact = LinearLogitArtifact::schema_canary_v0(
            MlTask::SettlementProbabilityResidual,
            "features-v1",
            ["distance"],
        )
        .unwrap();
        let mut json: serde_json::Value =
            serde_json::from_str(&artifact.to_json().unwrap()).unwrap();
        json["artifact_sha256"] = serde_json::Value::String("0".repeat(64));
        let error = LinearLogitArtifact::from_json(&json.to_string()).unwrap_err();
        assert!(matches!(error, ArtifactError::ArtifactHashMismatch { .. }));
    }

    #[test]
    fn ml_b_tasks_are_supported_as_shadow_artifacts() {
        for task in [
            MlTask::FokFillProbability,
            MlTask::FokPostFillToxicityProbability,
        ] {
            LinearLogitArtifact::schema_canary_v0(task, "execution-v1", ["queue_ahead"])
                .unwrap()
                .validate()
                .unwrap();
        }
    }
}
